packages/engine/src/avatar/functions/autopilotFunctions.ts

```diff
@@ -0,0 +1,97 @@
+import { AvatarAnimationComponentType } from '../components/AvatarAnimationComponent'
+import { AvatarControllerComponent, AvatarControllerComponentType } from '../components/AvatarControllerComponent'
+
+const interactionGroups = getInteractionGroups(
+  CollisionGroups.Avatars,
+  CollisionGroups.Ground | CollisionGroups.Default
+)
+const raycastArgs = {
+  type: SceneQueryType.Closest,
+  origin: new Vector3(),
+  direction: new Vector3(),
+  maxDistance: 250,
+  groups: interactionGroups
+} as RaycastArgs
+
+export const autopilotSetPosition = (entity: Entity) => {
+  const avatarControllerComponent = getComponent(entity, AvatarControllerComponent)
+  if (avatarControllerComponent.gamepadLocalInput.lengthSq() > 0) return
+
+  const physicsWorld = Engine.instance.currentWorld.physicsWorld
+  const world = Engine.instance.currentWorld
+
+  const castedRay = Physics.castRayFromCamera(world.camera, world.pointerState.position, physicsWorld, raycastArgs)
+  const rayNormal = new Vector3(castedRay[0].normal.x, castedRay[0].normal.y, castedRay[0].normal.z)
+  if (!castedRay.length || !assessWalkability(entity, rayNormal)) return undefined
+
+  const autopilotPosition = castedRay[0].position
+  avatarControllerComponent.autopilotWalkpoint = autopilotPosition as Vector3
+
+  placeMarker(avatarControllerComponent, rayNormal)
+}
+
+export const ScaleFluctuate = (Marker: Object3D, sinOffset = 4, scaleMultiplier = 0.2, pulseSpeed = 10) => {
+  const scalePulse = scaleMultiplier * (sinOffset + Math.sin(pulseSpeed * Engine.instance.currentWorld.elapsedSeconds))
+  Marker.scale.set(scalePulse, 1, scalePulse)
+  Marker.updateMatrixWorld()
+}
+
+export class AutopilotMarker {
+  object = undefined as Object3D | undefined
+
+  initializeMarker = (currentScene: Scene) => {
+    const markerGeometry = new CylinderGeometry(0.175, 0.175, 0.025, 24, 1)
+    const material = new MeshBasicMaterial({ color: '#00E14E' })
+    this.object = new Mesh(markerGeometry, material)
+    currentScene.add(this.object)
+  }
+}
+
+export const markerInstance = new AutopilotMarker()
+
+export async function placeMarker(controller: AvatarControllerComponentType, rayNormal: Vector3) {
+  if (!controller.autopilotWalkpoint) return
+
+  if (!markerInstance.object) markerInstance.initializeMarker(Engine.instance.currentWorld.scene)
+
+  const autopilotMarkerObject = markerInstance.object as Object3D
+
+  autopilotMarkerObject.visible = true
+  autopilotMarkerObject.position.set(
+    controller.autopilotWalkpoint.x,
+    controller.autopilotWalkpoint.y,
+    controller.autopilotWalkpoint.z
+  )
+  const newRotation = new Quaternion().setFromUnitVectors(V_010, rayNormal)
+  autopilotMarkerObject.rotation.setFromQuaternion(newRotation)
+
+  const waitForAutopilot = () => {
+    const reached = (resolve) => {
+      if (!controller.autopilotWalkpoint) resolve()
+      else setTimeout((_) => reached(resolve), 250)
```

Comment:
we want to avoid promises and setTimeout usage. any async logic will happen outside the animation frame and will reduce performance.

if the intention here is to just wait for the autopilot movement to be completed, we should use a state reactor to respond to state changes